SAMPLES = ["SRR097977", "SRR098026"]


rule all:
    input:
        "outputs/calls/all.vcf"


rule bwa_map:
    input:
        ref="rawdata/ecoli_rel606.fa",
        reads="rawdata/reads/{sample}.fastq"
    output:
        "outputs/mapped_reads/{sample}.bam"
    shell:
        "bwa mem {input.ref} {input.reads} | samtools view -Sb - > {output}"


rule samtools_sort:
    input:
        "outputs/mapped_reads/{sample}.bam"
    output:
        "outputs/sorted_reads/{sample}.bam"
    shell:
        "samtools sort -o {output} {input}"


rule samtools_index:
    input:
        "outputs/sorted_reads/{sample}.bam"
    output:
        "outputs/sorted_reads/{sample}.bam.bai"
    shell:
        "samtools index {input}"


rule bcftools_call:
    input:
        fa="rawdata/ecoli_rel606.fa",
        bam=expand("outputs/sorted_reads/{sample}.bam", sample=SAMPLES),
        bai=expand("outputs/sorted_reads/{sample}.bam.bai", sample=SAMPLES)
    output:
        "outputs/calls/all.vcf"
    shell:
        "samtools mpileup -g -f {input.fa} {input.bam} | "
        "bcftools call -mv - > {output}"
